backends/p4fmt/attach.cpp

```diff
@@ -0,0 +1,136 @@
+#include "attach.h"
+
+#include "ir/ir-generated.h"
+#include "lib/source_file.h"
+
+namespace P4 {
+
+void Attach::attachCommentsToNode(const IR::Node *node) {
+    const auto &comments = node->srcInfo.getAllComments();
+
+    if ((node == nullptr) || !node->srcInfo.isValid() || comments.empty()) {
+        return;
+    }
+
+    static std::unordered_set<const Util::Comment *> attachedComments;
+
+    const auto &nodePosition = node->srcInfo.getStart();
+    unsigned nodeLineNumber = nodePosition.getLineNumber();
+
+    for (auto *comment : comments) {
+        // Prevent attachment of comments to multiple nearest nodes
+        if (attachedComments.find(comment) != attachedComments.end()) {
+            // Comment is already attached to another node
+            continue;
+        }
+
+        const auto &commentStart = comment->getStartPosition();
+        const auto &commentEnd = comment->getEndPosition();
+
+        if (commentEnd.getLineNumber() == nodeLineNumber - 1) {
+            node->srcInfo.addBefore(comment);
+            attachedComments.insert(comment);
+        } else if (commentStart.getLineNumber() == nodeLineNumber) {
+            node->srcInfo.addAfter(comment);
+            attachedComments.insert(comment);
+        }
```

Comment:
I doubt this is the correct way to attach trailing inline comments. According to the [blog post](https://jayconrod.com/posts/129/preserving-comments-when-parsing-and-formatting-code) you shared, this should be done in a postorder traversal pass.

Reply:
Yes. I'm reworking that.